// ports/lib/port.js
exports.create = function createPort (setter, getter) {
    function port (val) {
        if (arguments.length == 0)
            return getter();
        setter(val);
    }

    port.__is_port = true;
    
    return port;
}

exports.createListenable = function createListenable (setter, getter, events) {
    var port = exports.create(setter, getter);

    port.listen = function (cb) {
        events.on('change', cb, port);
    }

    port.removeListener = function (cb) {
        events.off('change', cb);
    }

    port.removeListeners = function () {
        events.off('change');
    }

    return port;
}

exports.isPort = function (obj) {
    return !!obj.__is_port;
}

exports.link = function (src, target) {
    src.listen(function () {
        target(src());
    });
    target(src());
}

exports.linkTwoWay = function (src, target) {
    var updating = false;

    function updater (src, target) {
        return function () {
            if (updating) return;
            updating = true;
            target(src());
            updating = false;
        }
    }

    target(src());

    src.listen(updater(src, target));

    target.listen(updater(target, src));
}
