// client/pages/advert-edit/boot.js
var P = require('ports');
var Model = require('./model');
var Field = require('./field');
var Range = require('./range-field');

module.exports = function init (p) {
    window.F = new Form;
    
    $('*', p.el).ports(window.F, null, filters);

    p._show();   
}

function Form () {
    var self = this;
    this.rent = new Advert;
    this.sale = new Advert;

    this.hasError = P.computable(function () {
        return Model.hasError.call(self);
    }, this.rent.hasError, this.sale.hasError);
}

Form.prototype = Object.create(Model);


function Advert () {
    var self = this;
    this.available = Field(false);
    this.space = Field().mix(Range);
    this.price = Field().mix(Range);
    this.hasError = P.computable(function () {
        return self.available() && Model.hasError.call(self);
    }, this.space.hasError, this.price.hasError, this.available);
}

Advert.prototype = Object.create(Model);

var filters = {
    range: P.conversionExit(
        null,
        function output (val) {
            return require('range').toString(val);
        }
    )
}
