// client/node_modules/page.js

module.exports = Page;

function Page (el, url) {
    this.el = el;
    this.url = url;
    this._destroyed = false;
}

Page.prototype.load = function () {
    var p = this;
    this.$().load(p.url, function () {
        if (p._destroyed) return;
        p.init();
        if (!p._hidden) 
            p._show();
    });
}

Page.prototype.init = function () {
    
}

Page.prototype.close = function () {
    var p = this;
    p._hidden = true;
    p._hide();
    var isNotSuspended = !p.onsuspend();
    if (isNotSuspended) {
        p.ondestroy();
        p._destroyed = true;
        p.$().remove();
    }
    return p._destroyed;
}

Page.prototype.resume = function () {
    this.onresume();
    this._hidden = false;
    this._show();
}

Page.prototype._show = function () {
    this.$().show();
}

Page.prototype._hide = function () {
    this.$().hide(); 
}

Page.prototype.$ = function (selector) {
    return selector ? $(selector, this.el) : $(this.el);
}

Page.prototype.onsuspend = function () {
    return false;
}

Page.prototype.onresume = function () {
}

Page.prototype.ondestroy = function () {
}
